### ISSUE TITLE
Better support for automatic port allocation

### ISSUE THREAD

**Issue by allisonkarlitskaya** [ronf/asyncssh#603]
First: thanks for all of the work on this library.  I've just discovered it today, but from what I can see, it's amazingly comprehensive and seems to be actively developed with good modern practices.

## The issue
I want to use asyncssh to implement a server for unit testing.  I don't care which port number is used for the server, except that it needs to be unique.  Many tests may be running in parallel, so each test needs to get its own port number.  I'd like the OS to choose for me, which is normally possible by passing `0` as the port number to `bind()`.

It's possible to give a port number of `0` to `asyncssh.listen()` which will allocate a free port, but this seems to work almost by accident.  There are two issues:

 1. no attempt is made to allocate a single port for both IPv6 and IPv4: each gets its own unique port number.  That's https://github.com/python/cpython/issues/89856 in action.
 2. there's no good API for determining the chosen port(s) after the fact

### Try to inspect the sockets after the fact

The `SSHAcceptor` class implements a `__getattr__` to forward unknown attributes through to the underlying `asyncio.Server` object, and there we can inspect `sockets`.  This might look like so:

`(<asyncio.TransportSocket fd=6, family=10, type=1, proto=6, laddr=('::1', 40649, 0, 0)>, <asyncio.TransportSocket fd=7, family=2, type=1, proto=6, laddr=('127.0.0.1', 47331)>)
`

for a case where we did something like

```python3
    await asyncssh.listen('localhost', 0, ...)
```

We can go ahead iterating the list of sockets and check their listener addresses.  This works, but it's a bit clumsy.

### Bind your own port

You can also just create, `bind()`, and `listen()` your own socket, check the port number you got with `.getsockname()`, and then pass it as `sock=` to `asyncssh.listen()`.  That's probably the cleanest approach, but it feels a bit manual.

## Request

Given that the "two separate ports" issue is something that comes from Python itself, I'd understand if you don't want to try fixing it.  asyncssh could try to bind its own socket to avoid the issue, but as a user, I could just as well say `127.0.0.1` or `::1` depending on which one I want, and then I'd get a single port allocated.  What's needed here is clear documentation on what is to be expected when passing `0` or `None` as the port, particularly in the case that the given hostname resolves to more than one possible address.

The second ask is a nice API for determining the chosen port, after listening on it.  `SSHAcceptor` would be a good place to have a `get_port()` function which could return a simple `int` in case of a single listening socket (or throw a descriptive error in case of multiple sockets).  There could be `get_addresses()` or so for the more generic case.

If asyncssh tried to ensure a single port number then it might even be possible to simply set a `SSHAcceptor.port` int attribute, unconditionally.

**Comment by ronf** [comment]
Interestingly, there's already code in AsyncSSH which does something like what you're asking for, in `asyncssh.listener.create_tcp_local_listener()`. This function will creating multiple listening sockets (based on what getaddrinfo() returns as the resolved addresses), but it tries to make sure that all of the listeners end up on the same port. This can fail since the port choice for the first IP in the list may already be in use on the later IPs, but if it succeeds you'll end a collection of listening sockets all on the same port. I needed this to deal with the limitation that when you set up SSH port forwarding you can only return a single port number via the SSH protocol.

As a result of this limitiation, `create_tcp_forward_listener()` and `create_socks_listener()` use `asyncssh.listener.create_tcp_local_listener()` instead of the normal Python socket listener which can end up choosing different ports for each listening IP address, as you discovered.

There are some places in AsyncSSH such as the top-level `asyncssh.listen()` method where the standard Python `loop.create_server()` is called, and I just pass arguments through to that, letting it behave as much as possible like a traditional TCP socket would in Python, including the behavior you get when passing in a port number of 0. Those are the places you're seeing AsyncSSH potentially return multiple ports.

I could look into adding a new boolean argument on `asyncssh.listen()` like `single_port' which when set to `True` could try to open all sockets on a single port. It would have the same issue that `create_tcp_local_listener()` does today where that sometimes fails, though. Also, I'll need to look more closely at the existing `loop.create_server()` to see if there's a way to pass in a collection of already-bound listening sockets already listening on a chosen port. Based on an initial read, this might not be straightforward -- it looks like you can only pass in a single socket via the existing arguments.

Since I was originally just passing through all the socket-related arguments, I didn't feel it necessary to document the exact behavior. In fact, this may not be a good idea as the behavior could change over time as Python changes. If I addedthis new argument, I could document that and how AsyncSSH works when it binds the sockets. However, even there the behavior will be largely dependent on Python's `getaddrinfo()`, which I wouldn't be changing here.

As for a `get_port()` method, I like your suggestion of putting this on `SSHAcceptor`. Such a function already exists on the `SSHListener()` class, but I intentionally left it out in the case of `SSHAcceptor` because of the Python behavior or potentially binding to different ports. I could add it here in the case where `single_port` is `True`, or perhaps even allow it to work in other cases like when there's only a single address to bind to, but in the case where Python did bind to multiple ports, I'd probably end up needing `get_port()` to return something like `0` to indicate this. For `get_addresses()`, I'd probably look at returning a list of `getsockname()` results, so that would contain both addresses and ports.

Unfortunately, I'm not sure there's a good solution here which doesn't involve changes to `loop.create_server()`, at which point it would probably make more sense to have the new `single_port` option be an argument directly to `loop.create_server()` itself rather than `asyncssh.listen()`.

**Comment by allisonkarlitskaya** [comment]
I've never been a fan of the approach of creating two separate sockets and trying to bind them on a single port and hoping for the best.  I far prefer to create a single IPv6 socket with the `IPV6_ONLY` sockopt disabled and let that one pick a port.  Then it will accept on both IPv6 and IPv4.

**Comment by ronf** [comment]
I mostly agree, but I believe the reason this is done is that you can pass in a list of specific addresses to bind to (multiple IPv4 or IPv6), and that case absolutely requires multiple sockets. For the specific case of INADDR_ANY, a single socket could be used for both, but then you end up getting back addresses such as '::ffff:127.0.0.1' from `getsockname()` instead of just '127.0.0.1'. Since multiple sockets are needed to handle listening on multiple specific IPs, using them in this case avoids having to convert returned addresses from IPv6 to IPv4. Also, if you specify an empty string or `None` as the host, the stack automatically returns multiple entries (separate for each family) in `getaddrinfo()`. For instance:

```python
ocket.getaddrinfo(None, 0, family=socket.AF_UNSPEC, type=socket.SOCK_STREAM, flags=socket.AI_PASSIVE)
```

returns:

```python
[(<AddressFamily.AF_INET6: 30>, <SocketKind.SOCK_STREAM: 1>, 6, '', ('::', 0, 0, 0)),
 (<AddressFamily.AF_INET: 2>, <SocketKind.SOCK_STREAM: 1>, 6, '', ('0.0.0.0', 0))]
```

As long as IPV6_V6ONLY is set when opening the listening sockets, these entries can both be bound without it being a special case, and then you get back clean IPv4 or IPv6 addresses when you call `getsockname()`.

**Comment by allisonkarlitskaya** [comment]
I can kinda see the point here, and particularly the thing about "silly looking peer address returned from the socket" is something the [Python people specifically thought about as well](https://github.com/python/cpython/issues/80389#issuecomment-1093817363).

I also agree that the "please try to bind all sockets to the same port" functionality really belongs in Python itself.  I guess all that leaves here, then, is the "get port" API?

**Comment by ronf** [comment]
Yeah - I can still look into offering `get_addresses()` and `get_port()` in `SSHAcceptor`. It may take a week or two before I can start on this, though, as I'm right in the middle of another set of changes.

Also, there's one other idea I wanted to mention. It's not a perfect solution as there's a potential race condition in it, but when you are trying to listen on multiple IP addresses there's always going to be a risk of any port you get back from the stack already being in use by some of the IPs. Given that, I thought of the following:

```python
    sock = socket.socket()
    sock.bind(('', 0))
    port = sock.getsockname()[1]
    sock.close()

    print(f'Listening on port {port}...')

    await asyncssh.listen('', port, server_host_keys=['ssh_host_key'],
                          authorized_client_keys='ssh_user_ca',
                          line_editor=False, process_factory=handle_client)
```

Basically, you use the stack to find a random open port but then immediately close the socket associated with that port and pass it in as the port argument in `asyncssh.listen()`. It will then try to bind to that port on all of the IPs associated with the hostname you pass in (or localhost IPs if you pass in an empty string or `None`). This isn't something I'd want to include in AsyncSSH itself, but it's only a few lines of extra code to drop into your application to get what you were looking for. You might need to put a loop around it to try a few times if the port selected is grabbed by something else before AsyncSSH can listen on it, but pretty much any solution would potentially have such a race if trying to bind on multiple addresses.

From what I can tell, passing in the empty string when doing this initial `bind()` actually guarantees the port is not in use on _any_ IP address in the system at the time the port is selected. So, unless some other process comes along between the `sock.close()` and the `asyncssh.listen()`, this code should succeed in getting you the same port on all of the sockets it opens.

What do you think?

**Comment by allisonkarlitskaya** [comment]
I think I don't like hacks like this.  It's racy and it's also a bit unpythonic in my opinion (this pattern of check-first-and-then-do instead of just-do-and-deal-with-the-consequences).

For use in my testcases I'm just gonna bind to one interface (v6 or v4) or (less likely) use the non-v6only socket.

For the record, if I had to do a multi-socket approach with automatic port allocation, I'd do this one:

1. allocate the socket for the first address with port 0, see what the kernel gives us
2. try to use that kernel-assigned port with all of the other sockets
3. if it fails, close all the sockets and try again, up to 10 times or something

**Comment by ronf** [comment]
I agree that having to close the initial socket and then basically having to grab the same port again isn't ideal. However, binding on the empty string is the only way to ask the system to assign a port number that's not in use on ANY address, increasing your chance of the later bind() calls for all the individual addresses succeeding without having to retry.

The multi-socket approach you described is similar to what I did in `create_tcp_local_listener()`. However, that approach won't work today if your goal is to eventually do an `asyncio.create_server()`, since there's no way to pass more than one already-opened socket to that function. Changes would be need in the asyncio module if you wanted to take this approach. Also, as noted above, the kernel may select a port which is already in use on one or more of the later addresses you're going to try to bind to. This won't happen with the empty-string bind(), so the only chance of collision is if something else in the system grabs the selected port in between the call to `sock.close()` and `asyncio.create_server()`, which is highly unlikely.

Going with just a single listening socket with V6ONLY disabled is definitely an option if you aren't trying listen on specific IP addresses, and then you won't have any of these issues. However, if you want to listen on specific ports, disabling V6ONLY isn't an option.

Also, if you do go with disabling V6ONLY, don't forget to test the return value from `getsockname()` and/or `getpeername()` to see if it starts with `'::ffff:'`. If so, you'll probably want to strip that off.

**Comment by allisonkarlitskaya** [comment]
> Going with just a single listening socket with V6ONLY disabled is definitely an option if you aren't trying listen on specific IP addresses, and then you won't have any of these issues. However, if you want to listen on specific ports, disabling V6ONLY isn't an option.

There's one exception to this, and it's the most important one: listening on `::1` also allows clients to connect to you via `127.0.0.1`.

**Comment by ronf** [comment]
Yeah - listening on localhost definitely falls into the multiple-socket case. If you do the empty string bind() first to get the port, you can then proceed to let asyncio.create_server() bind the two separate sockets for you (both on the same port), but I don't see a way to do this without opening and closing an extra socket to get the stack to do the random port selection.

**Comment by ronf** [comment]
I've just added commit 71abf74 to the "develop" branch, with new `get_addresses()` and `get_port()` methods. If you get a chance to try it out, let me know if you see any problems.

**Comment by allisonkarlitskaya** [comment]
I tried my hand at an implementation and it's scary how close yours ended up being to mine, including using comprehensions and sets.

The feature, as implemented, looks like exactly what I need, thanks.  I'll give something like a review on the patch.

I didn't implement the check for `get_port()` existing on the server object.  You also  seem to have a stray debugging statement there:
```python3
    print(type(self._server))
```
Is the purpose of this code to detect a future `.get_port()` function which might get added?  I'd have probably approached that from the "forgiveness, not permission" standpoint and caught `AttributeError` had it been raised.

One other note: the reason I never finished my my implementation is that I got stuck on the worry that not all possible sockets that might be listened on will return a tuple with a port number as the second item.  It's possible, for example, to create a listener on a unix socket using `sock=` and then `.get_port()` will badly break.

Otherwise, I consider this issue adequately addressed, given the constraints on the standard library side.  I would be OK if this issue were closed now.  Thanks for the quick progress on this!

**Comment by ronf** [comment]
Thanks for catching the debugging line! I'll clean that up shortly. I meant to remove it before checking in.

The check for whether get_port() exists was necessary, as there are cases when setting up an SSH listener by tunneling over an existing SSH connection where the SSHAcceptor class ends up wrapping an SSHTCPClienterListener, rather than an asyncio.Server. In that case, there is no "sockets" member, and so the new get_addresses() and get_port() functions actually end up failing without this check.

For now, the implementation only fixes get_port() -- calling get_addresses() on a tunneled SSH listener will still break. I may try to fix that, but the best I can do in that case is to report back the listen host as-is for get_addresses(), as a single entry. It would still be a hostname in this case when one is provided, since the DNS lookup doesn't happen on the client. Instead, it happens on the remote SSH server the client tunnels through, and that only reports back a port number, not the resolved addresses it ends up listening to.

I hadn't really considered the case where you pass in a UNIX domain socket to listen(), since right now the primary API only allows for TCP listeners. That might be another case where the best we can do is to report 0 for the port, but since the caller knows they are passing in a UNIX domain socket, they should know that there's no point in calling get_port().

**Comment by ronf** [comment]
Ok - commit 4d8f2a4 adds code to allow get_addresses() to work in the tunneling case.

I started to put in changes to handle the UNIX domain socket case better, but it turns out that there are other changes needed in the unit test code before a UNIX domain socket can be passed in as a listener, so I think I'll hold off on this change for now. If I do add support for this, it probably should be done as a separate commit.

**Comment by ronf** [comment]
This feature is now available in AsyncSSH 2.14.1.
